src/data/upsert.ts

```diff
@@ -14,34 +11,81 @@ export const SparseValuesSchema = Type.Object(
   { additionalProperties: false }
 );
 
-const Vector = Type.Object(
+const VectorSchema = Type.Object(
   {
-    id: nonemptyString,
+    id: Type.String({ minLength: 1 }),
     values: Type.Array(Type.Number()),
     sparseValues: Type.Optional(SparseValuesSchema),
     metadata: Type.Optional(Type.Object({}, { additionalProperties: true })),
   },
   { additionalProperties: false }
 );
+const VectorArraySchema = Type.Array(VectorSchema);
 
-const VectorArray = Type.Array(Vector);
+const ChunkedUpsertSchema = Type.Object({
+  vectors: VectorArraySchema,
+  chunkSize: Type.Number({ minimum: 1 }),
+});
 
-export type Vector = Static<typeof Vector>;
+const UpsertOptionsSchema = Type.Union([
+  VectorArraySchema,
+  ChunkedUpsertSchema,
+]);
+
+export type Vector = Static<typeof VectorSchema>;
 export type SparseValues = Static<typeof SparseValuesSchema>;
-export type VectorArray = Static<typeof VectorArray>;
+export type VectorArray = Static<typeof VectorArraySchema>;
+export type UpsertOptions = Static<typeof UpsertOptionsSchema>;
 
 export const upsert = (api: VectorOperationsApi, namespace: string) => {
-  const validator = buildConfigValidator(VectorArray, 'upsert');
+  const validator = buildConfigValidator(UpsertOptionsSchema, 'upsert');
+
+  return async (options: UpsertOptions): Promise<void> => {
+    validator(options);
 
-  return async (vectors: VectorArray): Promise<void> => {
-    validator(vectors);
+    const isChunkedUpsert = !Array.isArray(options);
 
     try {
-      await api.upsert({ upsertRequest: { vectors, namespace } });
+      if (isChunkedUpsert) {
+        await chunkedUpsert(api, options.vectors, namespace, options.chunkSize);
+      } else {
+        await api.upsert({ upsertRequest: { vectors: options, namespace } });
+      }
       return;
     } catch (e) {
       const err = await handleDataError(e);
       throw err;
     }
   };
 };
+
+const chunkedUpsert = async (
+  api: VectorOperationsApi,
+  vectors: VectorArray,
+  namespace: string,
+  chunkSize = 10
+) => {
+  // Split vectors into chunks and promises for upserting
+  const chunks = sliceArrayToChunks(vectors, chunkSize);
+
+  const chunkedPromises = chunks.map(async (chunk) => {
+    try {
+      await api.upsert({ upsertRequest: { vectors: chunk, namespace } });
+    } catch (e) {
+      throw new Error(`Error upserting chunk: ${e}`);
```

Comment:
You will run into this when you try to rebase over the init changes. But I think you could use the `handleApiError` function here that I just merged in to standardize the error handling.

```
const err = await handleApiError(e);
throw err;
```

If you want to prefix the message with extra info you could do something like this by passing a second param.

```
const err = await handleApiError(e, async (statusCode, rawMessage) => (`Error upserting chunk. ${rawMessage}`));
throw err;
```